python/ray/air/execution/resources/placement_group.py

```diff
@@ -0,0 +1,173 @@
+import time
+from collections import defaultdict
+from typing import Dict, Optional, List, Set, Union
+
+from dataclasses import dataclass
+
+import ray
+from ray.air.execution.resources.request import (
+    ResourceRequest,
+    AllocatedResource,
+)
+from ray.air.execution.resources.resource_manager import ResourceManager
+from ray.util.annotations import DeveloperAPI
+from ray.util.placement_group import PlacementGroup, remove_placement_group
+from ray.util.scheduling_strategies import PlacementGroupSchedulingStrategy
+
+
+@DeveloperAPI
+@dataclass
+class PlacementGroupAllocatedResource(AllocatedResource):
+    placement_group: PlacementGroup
+
+    def annotate_remote_objects(
+        self, objects
+    ) -> List[Union[ray.ObjectRef, ray.actor.ActorHandle]]:
+        # With an empty head, the second bundle should live in the
+        # actual PG's first bundle, so we start counting from -1
+        if self.resource_request.head_bundle_is_empty:
+            start = -1
+            bundles = [{}] + self.resource_request.bundles
+        else:
+            start = 0
+            bundles = self.resource_request.bundles
+
+        annotated = []
+        for i, (obj, bundle) in enumerate(zip(objects, bundles), start=start):
+            bundle = bundle.copy()
+            num_cpus = bundle.pop("CPU", 0)
+            num_gpus = bundle.pop("GPU", 0)
+            memory = bundle.pop("memory", 0.0)
+
+            annotated.append(
+                obj.options(
+                    scheduling_strategy=PlacementGroupSchedulingStrategy(
+                        placement_group=self.placement_group,
+                        # Max ensures that empty head bundles are correctly placed
+                        placement_group_bundle_index=max(0, i),
+                        placement_group_capture_child_tasks=True,
+                    ),
+                    num_cpus=num_cpus,
+                    num_gpus=num_gpus,
+                    memory=memory,
+                    resources=bundle,
+                )
+            )
+        return annotated
+
+
+@DeveloperAPI
+class PlacementGroupResourceManager(ResourceManager):
+    _resource_cls: AllocatedResource = PlacementGroupAllocatedResource
+
+    def __init__(self, update_interval: float = 0.1):
+        self._pg_to_request: Dict[PlacementGroup, ResourceRequest] = {}
+        self._request_to_staged_pgs: Dict[
+            ResourceRequest, Set[PlacementGroup]
+        ] = defaultdict(set)
+        self._request_to_ready_pgs: Dict[
+            ResourceRequest, Set[PlacementGroup]
+        ] = defaultdict(set)
+
+        self._staging_future_to_pg: Dict[ray.ObjectRef, PlacementGroup] = dict()
+        self._pg_to_staging_future: Dict[PlacementGroup, ray.ObjectRef] = dict()
+        self._acquired_pgs: Set[PlacementGroup] = set()
+
+        self._update_interval = update_interval
+        self._last_update = time.monotonic() - self._update_interval - 1
```

Comment:
Because `time.monotonic()` does not have any guarantees around the value it emits, only that two sequential calls are monotonically increasing:

```
The reference point of the returned value is undefined, so that only the difference between the results of two calls is valid.
```

I guess `float("-inf")` would be an alternative